avr-hal-generic/src/i2c.rs

```diff
@@ -570,5 +570,16 @@ macro_rules! impl_twi_i2c {
                 }
             }
         }
+
+        impl<CLOCK, M> [<$I2c Slave>]<CLOCK, M>
+        where
+            CLOCK: $crate::clock::Clock,
+        {
+            pub fn start(&mut self) {
+                // this does not compile
+                // self.p.$twar.write(|w| unsafe {w.bits(self.address)});
```

Comment:
Does not compile because the macro token `$twar` is not defined anywhere.  You'll need to add it in the macro pattern , somewhere e.g.:

```diff
 registers: {
     ...
     bitrate: $twbr:ident,
+    address: $twar:ident,
     data: $twdr:ident,
 }
```

And of course add this to the macro invocations sites.

Reply:
Did I delete your comments? If so, sorry! Re-pasting back in here.

----
From Rahix
----

In avr-hal-generic/src/i2c.rs:

> @@ -570,5 +570,16 @@ macro_rules! impl_twi_i2c {
                 }
             }
         }
+
+        impl<CLOCK, M> [<$I2c Slave>]<CLOCK, M>
+        where
+            CLOCK: $crate::clock::Clock,
+        {
+            pub fn start(&mut self) {
+                // this does not compile
+                // self.p.$twar.write(|w| unsafe {w.bits(self.address)});

let's go through the line step by step:

    p is the peripheral. It's type is passed in from the macro-invocation and e.g. for ATmega328P will end up being avr_device::atmega328p::TWI. This type "derefs" to &avr_device::atmega328p::twi::RegisterBlock which is where fields for all the registers are found.
    $twsr is a macro token that the macro-invocation gives a proper register name for. This exists because the registers might have different names on different MCUs. For ATmega328P, it will just be twsr. The same story with the $twps field name for a bit-field in the register. The line will end up being expanded to

    p.twsr.write(|w| w.twps().prescaler_1()); 

    for ATmega328P.
    The .write() method is part of the svd2rust peripheral API. Read more about it in its documentation. Essentially, this line says: Write the whole register to its default values and set the prescaler bits to the bit-pattern for PRESCALER1.
    w.$twps() means "take the TWPS field in this register"
    and $twps().prescaler_1() means "write the bit-pattern for PRESCALER1 to it". This is also part of the autogenerated API (link 1, link 2).

Hope this helps ...

Reply:
Wow, didn't even know it was possible to delete other peoples comments :D

Reply:
So here's what I see so far...

1. `peripheral` -> `crate::pac::TWI` -> `RegisterBlock`
1. `RegisterBlock` contains all the TWI thingies, like `TWSR`

And in this generated file `avr-device/src/devices/atmega328p/twi.rs`, the TWI thingies are defined, like this:

```rust
pub type TWSR = crate::Reg<u8, _TWSR>;
#[allow(missing_docs)]
#[doc(hidden)]
pub struct _TWSR;
#[doc = "`read()` method returns [twsr::R](twsr::R) reader structure"]
impl crate::Readable for TWSR {}
#[doc = "`write(|w| ..)` method takes [twsr::W](twsr::W) writer structure"]
impl crate::Writable for TWSR {}
#[doc = "TWI Status Register"]
pub mod twsr;
```

And the part that I was missing was the implementation of the `crate::Readable` and `crate::Writeable`.

I think I'm starting to see how this is wired together now.

Reply:
I'm still a little confused as to this wiring...how does `crate::pac::TWI` deref to become `RegisterBlock`?

Yeah, the message that popped up was, 'the repository owner has deleted a comment'. I thought I had double posted a comment and was deleting the duplicate. But it turned out to be your comment. Totally weird that I could do that.

Reply:
> how does crate::pac::TWI deref to become RegisterBlock?

There's an

```rust
impl Deref for TWI {
	type Target = twi::RegisterBlock;

	...
}
```

somewhere in the auto-generated code.